File: day11/q2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <cmath>

using namespace std;

vector<long long> splitNumbers(const vector<long long>& numbers) {
    vector<long long> newNumbers;
    for (long long number : numbers) {
        if (number == 0) {
            newNumbers.push_back(1);
        } else {
            string numberStr = to_string(number);
            size_t length = numberStr.length();
            if (length % 2 == 0) {
                newNumbers.push_back(stoi(numberStr.substr(0, length / 2)));
                newNumbers.push_back(stoi(numberStr.substr(length / 2)));
            } else {
                newNumbers.push_back(number * 2024);
            }
        }
    }
    return newNumbers;
}

int main() {
    ifstream inputFile("input.txt");
    string line;

    if (!inputFile) {
        cerr << "Error opening file." << endl;
        return 1;
    }

    getline(inputFile, line);
    inputFile.close();

    vector<long long> numbers;
    stringstream ss(line);
    string temp;

    while (ss >> temp) {
        numbers.push_back(stoi(temp));
    }

    cout << "Initial numbers: ";
    for (long long num : numbers) {
        cout << num << " ";
    }
    cout << endl;

    for (long long i = 1; i <= 75; ++i) {
        numbers = splitNumbers(numbers);

        if (i < 7) {
            cout << "Iteration " << i << ": ";
            for (long long num : numbers) {
                cout << num << " ";
            }
            cout << endl;
        }

        cout << i << ": " << numbers.size() << endl;
    }

    return 0;
}

File: 2024/day11/q2.py
with open("input.txt") as f:
    line = f.readline()
numbers = line.split()
numbers_l = list(map(int, numbers))

numbers = {}
for number in numbers_l:
    if number not in numbers:
        numbers[number] = 0
    numbers[number] += 1

print(numbers)

for i in range(1, 76):
    new_numbers = {}
    for number, count in numbers.items():
        if number == 0:
            if 1 not in new_numbers:
                new_numbers[1] = 0
            new_numbers[1] += numbers[0]
        elif len(str(number)) % 2 == 0:
            string = str(number)
            length = len(string)
            
            if int(string[:length//2]) not in new_numbers:
                new_numbers[int(string[:length//2])] = 0
            if int(string[length//2:]) not in new_numbers:
                new_numbers[int(string[length//2:])] = 0
            
            new_numbers[int(string[:length//2])] += count
            new_numbers[int(string[length//2:])] += count
        else:
            if number * 2024 not in new_numbers:
                new_numbers[number * 2024] = 0
            new_numbers[number * 2024] += count
            
    numbers = new_numbers
    
    if i < 7:
        print(numbers)
    sum = 0
    for number, count in numbers.items():
        sum += count
    print(f"{i}: {sum}")